test/integration/handlers/authenticate-handler.spec.ts

```diff
@@ -0,0 +1,710 @@
+      const handler = new AuthenticateHandler({
+        model: { getAccessToken() {} },
+      });
+      const request = new Request({
+        body: {},
+        headers: { Authorization: 'Bearer foo' },
+        method: 'ANY',
+        query: { access_token: 'foo' },
+      });
+
+      try {
+        handler.getTokenFromRequest(request);
+
+        should.fail('should.fail', '');
+      } catch (e) {
+        e.should.be.an.instanceOf(InvalidRequestError);
+        e.message.should.equal(
+          'Invalid request: only one authentication method is allowed',
+        );
+      }
+    });
+
+    it('should throw an error if `accessToken` is missing', () => {
+      const handler = new AuthenticateHandler({
+        model: { getAccessToken() {} },
+      });
+      const request = new Request({
+        body: {},
+        headers: {},
+        method: 'ANY',
+        query: {},
+      });
+
+      try {
+        handler.getTokenFromRequest(request);
+
+        should.fail('should.fail', '');
+      } catch (e) {
+        e.should.be.an.instanceOf(UnauthorizedRequestError);
+        e.message.should.equal('Unauthorized request: no authentication given');
+      }
+    });
+  });
+
+  describe('getTokenFromRequestHeader()', () => {
+    it('should throw an error if the token is malformed', () => {
+      const handler = new AuthenticateHandler({
+        model: { getAccessToken() {} },
+      });
+      const request = new Request({
+        body: {},
+        headers: {
+          Authorization: 'foobar',
+        },
+        method: 'ANY',
+        query: {},
+      });
+
+      try {
+        handler.getTokenFromRequestHeader(request);
+
+        should.fail('should.fail', '');
+      } catch (e) {
+        e.should.be.an.instanceOf(InvalidRequestError);
+        e.message.should.equal(
+          'Invalid request: malformed authorization header',
+        );
+      }
+    });
+
+    it('should return the bearer token', () => {
+      const handler = new AuthenticateHandler({
+        model: { getAccessToken() {} },
+      });
+      const request = new Request({
+        body: {},
+        headers: {
+          Authorization: 'Bearer foo',
+        },
+        method: 'ANY',
+        query: {},
+      });
+
+      const bearerToken = handler.getTokenFromRequestHeader(request);
+
+      bearerToken.should.equal('foo');
+    });
+  });
+
+  describe('getTokenFromRequestQuery()', () => {
+    it('should throw an error if the query contains a token', () => {
+      const handler = new AuthenticateHandler({
+        model: { getAccessToken() {} },
+      });
+
+      try {
+        handler.getTokenFromRequestQuery(undefined);
+
+        should.fail('should.fail', '');
+      } catch (e) {
+        e.should.be.an.instanceOf(InvalidRequestError);
+        e.message.should.equal(
+          'Invalid request: do not send bearer tokens in query URLs',
+        );
+      }
+    });
```

Comment:
This test isn't doing what it says: It doesn't fail because the query contains a token, it fails because the function `getTokenFromRequestQuery` is called without `allowBearerTokensInQueryString` being set to true.

I suspect this should be split into multiple cases with suitable descriptions: 
1. Check that this function fails when the property is not set/set to false
2. Check that it returns the token when the property is set to true (this case exists)

The case "fail when the token is set and the property isn't" should be added in the generic tests for `getTokenFromRequest` then: fail _if and only if_ that token exists in the query _and_ the property isn't set.